libraries/libopae-c/opae-cfg.c

```diff
@@ -0,0 +1,384 @@
+	// configuration table entries.
+	for (i = 0 ; i < num_plugins ; ++i) {
+		json_object *j_plugin_i =
+			json_object_array_get_idx(j_plugin, i);
+		int num_devices_i = 0;
+		json_object *j_devices = NULL;
+		json_object *j_enabled = NULL;
+		bool enabled = false;
+		json_object *j_module = NULL;
+		char *module = NULL;
+		json_object *j_configuration = NULL;
+		char *configuration = NULL;
+
+		j_enabled =
+			parse_json_boolean(j_plugin_i, "enabled", &enabled);
+		if (!j_enabled || !enabled) {
+			OPAE_DBG("plugin[%d] in %s not enabled. skipping",
+				 i, cfg_name);
+			continue;
+		}
+
+		j_devices =
+			parse_json_array(j_plugin_i, "devices", &num_devices_i);
+		if (!j_devices || !num_devices_i) {
+			OPAE_DBG("plugin[%d] \"devices\" in %s missing "
+				 "or empty. skipping", i, cfg_name);
+			continue;
+		}
+
+		j_module =
+			parse_json_string(j_plugin_i, "module", &module);
+		if (!j_module || !strlen(module)) {
+			OPAE_ERR("plugin[%d] \"module\" in %s missing "
+				 "or empty.", i, cfg_name);
+			res = 10;
+			goto out_free;
+		}
+
+		if (!json_object_object_get_ex(j_plugin_i,
+					       "configuration",
+					       &j_configuration)) {
+			OPAE_ERR("plugin[%d] \"configuration\" in %s missing "
+				 "or invalid.", i, cfg_name);
+			res = 11;
+			goto out_free;
+		}
+
+		configuration = (char *)
+			json_object_to_json_string_ext(j_configuration,
+						       JSON_C_TO_STRING_PLAIN);
+
+		for (j = 0 ; j < num_devices_i ; ++j) {
+			json_object *j_device_j =
+				json_object_array_get_idx(j_devices, j);
+			const char *dev_name = NULL;
+			libopae_config_data *pcfg;
+			opae_pci_device *dev;
+
+			if (!json_object_is_type(j_device_j,
+						 json_type_string)) {
+				OPAE_ERR("Non-string JSON item "
+					 "found in 'devices[%d]'", j);
+				res = 11;
+				goto out_free;
+			}
+
+			dev_name = json_object_get_string(j_device_j);
+
+			// Find the device named dev_name in pci_devices.
+			for (d = 0 ; d < num_devices ; ++d) {
+				if (!strncmp(dev_name,
+					     pci_devices[d].name,
+					     MAX_DEV_NAME))
+					break;
+			}
+
+			if (d == num_devices) {
+				OPAE_ERR("device %s not found\n", dev_name);
+				res = 12;
+				goto out_free;
+			}
+
+			pcfg = ctx->current;
+			++(ctx->current);
+			dev = &pci_devices[d];
+
+			pcfg->vendor_id = dev->vendor_id;
+			pcfg->device_id = dev->device_id;
+			pcfg->subsystem_vendor_id = dev->subsystem_vendor_id;
+			pcfg->subsystem_device_id = dev->subsystem_device_id;
+
+			pcfg->module_library = opae_strdup(module);
+			pcfg->config_json = opae_strdup(configuration);
+
+			if (!pcfg->module_library || !pcfg->config_json) {
+				OPAE_ERR("strdup() failed");
+				res = 13;
+				goto out_free;
+			}
+		}
+	}
+
+out_free:
+	opae_free(pci_devices);
+	return res;
+}
+
+libopae_config_data *
+opae_parse_libopae_json(const char *json_input)
+{
+	enum json_tokener_error j_err = json_tokener_success;
+	json_object *root = NULL;
+	json_object *j_configs = NULL;
+
```

Comment:
input validation

Reply:
Look at the call site.  json_input will be non-NULL.